// nexmo/lib/Nexmo.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _Credentials = require('./Credentials');

var _Credentials2 = _interopRequireDefault(_Credentials);

var _Message = require('./Message');

var _Message2 = _interopRequireDefault(_Message);

var _Voice = require('./Voice');

var _Voice2 = _interopRequireDefault(_Voice);

var _Number = require('./Number');

var _Number2 = _interopRequireDefault(_Number);

var _Verify = require('./Verify');

var _Verify2 = _interopRequireDefault(_Verify);

var _NumberInsight = require('./NumberInsight');

var _NumberInsight2 = _interopRequireDefault(_NumberInsight);

var _App = require('./App');

var _App2 = _interopRequireDefault(_App);

var _Account = require('./Account');

var _Account2 = _interopRequireDefault(_Account);

var _CallsResource = require('./CallsResource');

var _CallsResource2 = _interopRequireDefault(_CallsResource);

var _HttpClient = require('./HttpClient');

var _HttpClient2 = _interopRequireDefault(_HttpClient);

var _NullLogger = require('./NullLogger');

var _NullLogger2 = _interopRequireDefault(_NullLogger);

var _ConsoleLogger = require('./ConsoleLogger');

var _ConsoleLogger2 = _interopRequireDefault(_ConsoleLogger);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Nexmo =
/**
 * @param {Credentials} credentials - Nexmo API credentials
 * @param {string} credentials.apiKey - the Nexmo API key
 * @param {string} credentials.apiSecret - the Nexmo API secret
 * @param {Object} options - Additional options
 * @param {boolean} options.debug - `true` to turn on debug logging
 * @param {Object} options.logger - Set a custom logger.
 * @param {string} options.appendToUserAgent - A value to append to the user agent.
 *                    The value will be prefixed with a `/`
 */
function Nexmo(credentials) {
  var options = arguments.length <= 1 || arguments[1] === undefined ? { debug: false } : arguments[1];

  _classCallCheck(this, Nexmo);

  this.credentials = _Credentials2.default.parse(credentials);
  this.options = options;

  // If no logger has been supplied but debug has been set
  // default to using the ConsoleLogger
  if (!this.options.logger && this.options.debug) {
    this.options.logger = new _ConsoleLogger2.default();
  } else if (!this.options.logger) {
    // Swallow the logging
    this.options.logger = new _NullLogger2.default();
  }

  var userAgent = 'nexmo-node/UNKNOWN/UNKNOWN';
  try {
    var packageDetails = require(__dirname + '/../package.json');
    userAgent = 'nexmo-node/' + packageDetails.version + '/' + process.version;
  } catch (e) {
    console.warn('Could not load package details');
  }
  this.options.userAgent = userAgent;
  if (this.options.appendToUserAgent) {
    this.options.userAgent += '/' + this.options.appendToUserAgent;
  }
  this.options.httpClient = new _HttpClient2.default(this.options);

  this.message = new _Message2.default(this.credentials, this.options);
  this.voice = new _Voice2.default(this.credentials, this.options);
  this.number = new _Number2.default(this.credentials, this.options);
  this.verify = new _Verify2.default(this.credentials, this.options);
  this.numberInsight = new _NumberInsight2.default(this.credentials, this.options);
  this.applications = new _App2.default(this.credentials, this.options);
  this.account = new _Account2.default(this.credentials, this.options);
  this.calls = new _CallsResource2.default(this.credentials, this.options);

  /**
   * @deprecated Please use nexmo.applications
   */
  this.app = this.applications;
};

exports.default = Nexmo;
module.exports = exports['default'];
